src/content/en/fundamentals/engage-and-retain/push-notifications/common-notification-patterns.md

```diff
@@ -0,0 +1,375 @@
+
+The code to getting all the windows and looking for a focused window looks like this:
+
+    function isClientFocused() {
+      return clients.matchAll({
+        type: 'window',
+        includeUncontrolled: true
+      })
+      .then((windowClients) => {
+        let clientIsFocused = false;
+
+        for (let i = 0; i < windowClients.length; i++) {
+          const windowClient = windowClients[i];
+          if (windowClient.focused) {
+            clientIsFocused = true;
+            break;
+          }
+        }
+
+        return clientIsFocused;
+      });
+    }
+
+We use [clients.matchAll()](https://developer.mozilla.org/en-US/docs/Web/API/Clients/matchAll)
+to get all of our window clients and then we loop over them checking the `focused` parameter.
+
+Inside our push event we'd use this function to decide if we need to show a notification or not:
+
+      const promiseChain = isClientFocused()
+      .then((clientIsFocused) => {
+        if (clientIsFocused) {
+          console.log('Don\'t need to show a notification.');
+          return;
+
+        }
+
+        // Client isn't focused, we need to show a notification.
+        return self.registration.showNotification('Had to show a notification.');
+      });
+
+      event.waitUntil(promiseChain);
+
+## Message Page from a Push Event
+
+We've seen that you can skip showing a notification is the user is currently on your site. But
+what if you still wanted to let the user know that an event has occurred, but a notification is
+too heavy handed?
+
+One approach is to send a message from the service worker to the page, this way the web page
+can show a notification or update to the user informing them of the event. This is useful for
+scenarios when a subtle notification in the page is better / friendlier for the user.
+
+Let's say we've received and push and checked that our web app is currently focused, we can
+"post a message" to each open page, like so:
+
+      const promiseChain = isClientFocused()
+      .then((clientIsFocused) => {
+        if (clientIsFocused) {
+          windowClients.forEach((windowClient) => {
+            windowClient.postMessage({
+              message: 'Received a push message.',
+              time: new Date().toString()
+            });
+          });
+        } else {
+          return self.registration.showNotification('No focused windows', {
+            body: 'Had to show a notification instead of messaging each page.'
+          });
+        }
+      });
+
+      event.waitUntil(promiseChain);
+
+In each of pages, we listen for these messages by adding a message event
+listener:
+
+        navigator.serviceWorker.addEventListener('message', function(event) {
+          console.log('Received a message from service worker: ', event.data);
+        });
+
+In this message listener you could do anything you want, show a toast, a
+notification bubble or completely ignore the message.
```

Comment:
These items need links, images, or some other clarification. What's a toast? By notification bubble I assume you mean [`new Notification()`](https://developer.mozilla.org/en-US/docs/Web/API/Notification/Notification).